### ISSUE TITLE
Support cuPy array in custom objective with the sklearn interface.

### ISSUE THREAD

**Issue by trivialfis** [dmlc/xgboost#10049]
Currently, arrays for both prediction and labels are converted into `np.ndarray`. We can avoid the data transfer by:
- Output GPU prediction in the sklearn interface.
- Return GPU array in `DMatrix.get_float_info`
